From 54222fd144312f4e8b4fc2b28e9bd779b3054e92 Mon Sep 17 00:00:00 2001
From: Sunil Goutham <sgoutham@marvell.com>
Date: Fri, 8 Feb 2019 10:56:29 +0530
Subject: [PATCH 154/767] soc: octeontx2-rm: Defer probe if discovery id is not
 setup

commit b04dfe278aa44a13e96a5eaac28af62823d8859c from
git@git.assembla.com:cavium/WindRiver.linux.git

Check if discovery ID is setup by AF, if not, defer
driver probe until AF is up.

Change-Id: I0e72905389239e6342e09a75f1f73d2657cea29e
Signed-off-by: Sunil Goutham <sgoutham@marvell.com>
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/soc/marvell/octeontx2-rm/otx2_rm.c | 23 ++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/drivers/soc/marvell/octeontx2-rm/otx2_rm.c b/drivers/soc/marvell/octeontx2-rm/otx2_rm.c
index a398603a4319..29ed1cd899de 100644
--- a/drivers/soc/marvell/octeontx2-rm/otx2_rm.c
+++ b/drivers/soc/marvell/octeontx2-rm/otx2_rm.c
@@ -1085,6 +1085,25 @@ static int vf_sysfs_create(struct pci_dev *pdev)
 	return err;
 }
 
+static int rm_check_pf_usable(struct rm_dev *rm)
+{
+	u64 rev;
+
+	rev = rm_read64(rm, BLKADDR_RVUM, 0,
+			RVU_PF_BLOCK_ADDRX_DISC(BLKADDR_RVUM));
+	rev = (rev >> 12) & 0xFF;
+	/* Check if AF has setup revision for RVUM block,
+	 * otherwise this driver probe should be deferred
+	 * until AF driver comes up.
+	 */
+	if (!rev) {
+		dev_warn(&rm->pdev->dev,
+			 "AF is not initialized, deferring probe\n");
+		return -EPROBE_DEFER;
+	}
+	return 0;
+}
+
 static int rm_probe(struct pci_dev *pdev, const struct pci_device_id *id)
 {
 	struct device *dev = &pdev->dev;
@@ -1138,6 +1157,10 @@ static int rm_probe(struct pci_dev *pdev, const struct pci_device_id *id)
 		goto set_mask_failed;
 	}
 
+	err = rm_check_pf_usable(rm);
+	if (err)
+		goto set_mask_failed;
+
 	/* Map PF-AF mailbox memory */
 	rm->af_mbx_base = ioremap_wc(pci_resource_start(pdev, PCI_MBOX_BAR_NUM),
 				     pci_resource_len(pdev, PCI_MBOX_BAR_NUM));
-- 
2.31.1

